### ISSUE TITLE
light client: startup error Fork choice doesn't happen on light clients.

### ISSUE THREAD

**Issue by niklasad1** [paritytech/substrate#2541]
When I start the light client I get the following error:

```bash
➜  substrate git:(master) ✗ cargo run --release \-- --name niklasad1 --light
    Finished release [optimized] target(s) in 0.21s
     Running `target/release/substrate --name niklasad1 --light`
2019-05-10 23:05:35 Substrate Node
2019-05-10 23:05:35   version 2.0.0-6e67eff06-x86_64-linux-gnu
2019-05-10 23:05:35   by Parity Technologies, 2017-2019
2019-05-10 23:05:35 Chain specification: Emberic Elm
2019-05-10 23:05:35 Node name: niklasad1
2019-05-10 23:05:35 Roles: LIGHT
2019-05-10 23:05:35 Generated a new keypair: 592d1194c0ebc122b2c4a746da21f1188382fe82e701e9db4cc5a8b045a8ae81 (5E5dWN7T...)
2019-05-10 23:05:36 Initializing Genesis block/state (state: 0x44f3…c37f, header-hash: 0x10c0…d5c3)
Error: Error(Msg("Fork choice doesn\'t happen on light clients."), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })
```

**Comment by niklasad1** [comment]
Seems like there is an issue with the newly introduced `SelectChain` trait [#2240](https://github.com/paritytech/substrate/pull/2240/files). 

AFAIU https://github.com/paritytech/substrate/blob/master/core/service/src/lib.rs#L155 will always return `Err` in the `light client` and thus the `Service` will not get started!

/cc @gnunicorn Can you confirm or explain? Also, for the `light client` shall we refactor this or just provide a `dummy impl` for `SelectChain`?

**Comment by gnunicorn** [comment]
Sorry, yes, I forgot to resolve this or state this is happening now. `SelectChain` is not actually something that light clients are to be concerned with, thus it should (and is ) only be implemented for full-backends. But as the service attempt to be generic over light and full client creation, the API needed to be added for light clients, but - as of now - only fails to ever provide a select chain.

Providing a different dummy implementation won't work, because the service expected them to provide the same type for both clients, thus I opted for error'ing for now. Not sure what the proper solution should be (have `Option<SelectChain>`? Or maybe have a specific `NotOnLightClientError` that the code can handle directly -- maybe a bit more explicit but then the service must know about said context?) but I believe this won't work without a bigger refactor :( .

(an attempt to provide a dummy failed for the interfacing reasons mentioned)

**Comment by marcio-diaz** [comment]
I was trying the `Option<SelectChain>` approach on the weekend, thinking that could be straightforward, but I got stuck at some point, didn't know from where to take the `best header`. But I didn't spend much time.

**Comment by niklasad1** [comment]
Closed by #2574
